lib/AST/DiagnosticEngine.cpp

```diff
@@ -526,21 +526,30 @@ static void formatDiagnosticArgument(StringRef Modifier,
     Out << FormatOpts.ClosingQuotationMark;
     break;
 
+  case DiagnosticArgumentKind::FullyQualifiedType:
   case DiagnosticArgumentKind::Type: {
     assert(Modifier.empty() && "Improper modifier for Type argument");
     
     // Strip extraneous parentheses; they add no value.
-    auto type = Arg.getAsType()->getWithoutParens();
+    Type type;
+    bool needsQualification = false;
```

Comment:
While you are here, what do you think about enabling this by default for generic parameters? It's very common to name generic parameters like `T` or `U` which results in diagnostics that say something like "T is not convertible to T" and it would be much better to always qualify it to avoid confusion.

Reply:
I think something like a `DescriptiveTypeKind` would help out there too. 
```
T is not convertible to generic parameter T
```

Reply:
The problem is in most cases we don't know what type we'd get in conversion diagnostics so it'd be better to always qualify generic parameters...

Reply:
e.g.

```
protocol P {}
func foo<T: P>(_: T) {}
func bar<T>(t: T) {
  foo(t)
}
```

This would produce a note which says `where 'T' = 'T'`

Reply:
I think it would have been better if it did say `where 'T' = generic parameter bar.T` or something similar.

Reply:
Another example with shadowing:

```
protocol MyProtocol {}

struct Generic<M: MyProtocol> {
    var values: [M] = []

    func valueAt<M>(_ index: Int) -> M? {
        if values.indices.contains(index) {
            return values[index]
        } else {
            return nil
        }
    }
}
```

Currently results in `cannot convert return expression of type 'M' to return type 'M?'`

Reply:
Good point. I'll build the abstractions for this in a follow-up.

Reply:
Sounds great, thank you!